server/service/apple_mdm.go

```diff
@@ -3426,3 +3428,219 @@ func (svc *MDMAppleDDMService) handleDeclarationStatus(ctx context.Context, dm *
 
 	return nil
 }
+
+////////////////////////////////////////////////////////////////////////////////
+// Generate ABM keypair endpoint
+////////////////////////////////////////////////////////////////////////////////
+
+type generateABMKeypairResponse struct {
+	PublicKey []byte `json:"public_key,omitempty"`
+	Err       error  `json:"error,omitempty"`
+}
+
+func (r generateABMKeypairResponse) error() error { return r.Err }
+
+func generateABMKeypairEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (errorer, error) {
+	keyPair, err := svc.GenerateABMKeypair(ctx)
+	if err != nil {
+		return generateABMKeypairResponse{
+			Err: err,
+		}, nil
+	}
+
+	return generateABMKeypairResponse{
+		PublicKey: keyPair.PublicKey,
+	}, nil
+}
+
+func (svc *Service) GenerateABMKeypair(ctx context.Context) (*fleet.MDMAppleDEPKeyPair, error) {
+	if err := svc.authz.Authorize(ctx, &fleet.AppleBM{}, fleet.ActionWrite); err != nil {
+		return nil, err
+	}
+	var publicKeyPEM, privateKeyPEM []byte
+	assets, err := svc.ds.GetMDMConfigAssetsByName(ctx, []fleet.MDMAssetName{
```

Comment:
@mna sounds good, I will move it to the datastore! should we return a `notFound` error if there's one missing? this also makes me wonder if this datastore method shouldn't just return a single row and get called multiple times instead?

Reply:
I think the batch approach makes sense, we often need to read 2 or 3 values so it should be more efficient. Good question about using `notFound` or not, it would result in a 404 if the service layer returned a `notFound` directly, not sure that's what we want, right? Maybe a custom error in that case (return the map of found values AND the error)? Something like an `ErrPartialResult`? Seems like this could be a pattern we'd reuse for other batch stuff too.

Reply:
that's a great idea, thanks!